aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -599,6 +599,109 @@ Tensor& diff_out(const Tensor& self, int64_t n, int64_t dim, const c10::optional
+    TORCH_CHECK(self.device() == spacing[i].device(), "torch.gradient requires each tensor in spacing to be in the same device as self.");
+  }
+  pre_check_gradient(self, axis, spacing.size(), edge_order);
+
+  std::vector<Tensor> result;
+  for (const auto i : c10::irange(axis.size())) {
+    TORCH_CHECK( spacing[i].dim() < 2, "torch.gradient's argument spacing can receive eiter a scalar, a list of scalars or a list of one-dimensional tensors");
+    int64_t direction = maybe_wrap_dim(axis[i], self.dim());
+
+    if (spacing[i].dim()==1) {
+      std::vector<int64_t> shape(self.dim(),1);
+      shape[ direction ] = -1;
+
+      auto ax_dx = spacing[i].diff(1,0);
+      auto dx1 = at::slice(ax_dx, 0, 0, -1);
+      auto dx2 = at::slice(ax_dx, 0, 1);
+      auto a = (   -dx2    / (dx1*(dx1+dx2)) ).reshape(shape);
+      auto b = ( (dx2-dx1) / (dx1*dx2)       ).reshape(shape);
+      auto c = (    dx1    / (dx2*(dx1+dx2)) ).reshape(shape);
+
+      auto center  = a*at::slice(self, direction, 0, -2) + b*at::slice(self, direction , 1, -1) + c*at::slice(self, direction ,2);
+      auto prepend = (at::slice(self, direction, 1, 2  ) - at::slice(self, direction, 0, 1   )) / ax_dx[0]  ;
+      auto append  = (at::slice(self, direction, -1    ) - at::slice(self, direction, -2, -1 )) / ax_dx[-1] ;
+      result.push_back(prepend_append_on_dim(center, prepend, append, direction));
+    } else {
+      auto ax_dx = spacing[i];
+      auto center  = (at::slice(self,direction, 2   ) - at::slice(self, direction, 0, -2 ) ) / ax_dx;
+      auto prepend = (at::slice(self,direction, 1, 2) - at::slice(self, direction, 0, 1  ) ) / ax_dx;
+      auto append  = (at::slice(self,direction, -1  ) - at::slice(self, direction, -2, -1) ) / ax_dx;
+      result.push_back(prepend_append_on_dim(center/2, prepend, append, direction));
+    }
+  }
+  return result;
+}
+
+std::vector<Tensor> gradient_helper_scalar(const Tensor& self, ArrayRef<Scalar> spacing, IntArrayRef axis, int64_t edge_order) {
+  pre_check_gradient(self, axis, spacing.size(), edge_order);
+  std::vector<Tensor> result;
+  for (const auto i : c10::irange(axis.size())) {
+      int64_t direction = maybe_wrap_dim(axis[i], self.dim());
+      auto ax_dx = spacing[i];
+      auto center  = (at::slice(self,direction, 2   ) - at::slice(self, direction, 0, -2 ) ) / ax_dx;
+      auto prepend = (at::slice(self,direction, 1, 2) - at::slice(self, direction, 0, 1  ) ) / ax_dx  ;
+      auto append  = (at::slice(self,direction, -1  ) - at::slice(self, direction, -2, -1) ) / ax_dx ;
+      result.push_back(prepend_append_on_dim(center/2, prepend, append, direction));
+  }
+  return result;
+}
+
+std::vector<int64_t> gradient_axis_preprocess(const Tensor& self, c10::optional<int64_t> dim) {
+  std::vector<int64_t> axis;
+  if (!dim.has_value()) {
+    for (const auto i : c10::irange(self.dim())) {
+      axis.push_back(i);
+    }
+  }
+  else {
+    axis.push_back(dim.value());
```

Comment:
prefer emplace_back to push_back

Reply:
Actually I think function can be code golfed like this (although the way this function works might need to change):

```
if (dim.has_value()) {
  return std::vector<int64_t>{dim.value()};
}

std::vector<int64_t> axis(self.dim());
std::iota(axis.begin(), axis.end(), 0);
return axis;
```

The standard library is your friend.